subprojects/build-events/src/main/java/org/gradle/internal/build/event/BuildEventListenerRegistryInternal.java

```diff
@@ -18,20 +18,44 @@
 
 import org.gradle.api.provider.Provider;
 import org.gradle.build.event.BuildEventsListenerRegistry;
+import org.gradle.internal.code.UserCodeSource;
 import org.gradle.internal.operations.BuildOperationListener;
 import org.gradle.internal.service.scopes.Scope;
 import org.gradle.internal.service.scopes.ServiceScope;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
 
 import java.util.List;
 
-@ServiceScope(Scope.Global.class)
+@ServiceScope(Scope.BuildSession.class)
```

Comment:
💭 Need to reduce scope to get `UserCodeApplicationContext`. Ideally it should be limited to `Build`, but that is more involved.

Reply:
The application context is available in **cross**-build-session scope, so we probably should keep the scoping consistent: the scope of a service should normally be the lowest/deepest scope of its dependencies.